src/gpuarray_reduction.c

```diff
@@ -6,841 +6,4095 @@
+	const GpuReduction* gr;
+	GpuArray*           d0;
+	GpuArray*           d1;
+	const GpuArray*     s0;
+	int                 reduxLen;
+	const int*          reduxList;
+	int                 flags;
+
+	/* General. */
+	int                 nds0;         /* # Source                           axes */
+	int                 nds0r;        /* # Reduced                          axes */
+	int                 ndd0;         /* # Destination                      axes */
+	int                 ndfs0;        /* # Flattened source                 axes */
+	int                 ndib;         /* # Intra-block                      axes */
+	int                 zeroAllAxes;  /* # of zero-length                   axes in source tensor */
+	int                 zeroRdxAxes;  /* # of zero-length         reduction axes in source tensor */
+	size_t              prodAllAxes;  /* Product of length of all           axes in source tensor */
+	size_t              prodRdxAxes;  /* Product of length of all reduction axes in source tensor */
+	size_t              prodFreeAxes; /* Product of length of all free      axes in source tensor */
+	
+	/* Flattening */
+	axis_desc*          xdSrc;
+	axis_desc**         xdSrcPtrs;
+	axis_desc*          xdSplit;
+
+	/* Invoker */
+	uint32_t            selector;
+	uint64_t            U;
+	uint64_t            V;
+	uint64_t            B;
+	uint32_t            D;
+	uint32_t            Dunit;
+	uint32_t            H;
+	
+	uint32_t            LSlice;
+	uint64_t            LPadded;
+	uint64_t*           L;
+	uint32_t*           Li;
+	gpudata*            S0Data;
+	int64_t             S0Off;
+	int64_t*            S0J, *S0Si;
+	gpudata*            D0Data;
+	int64_t             D0Off;
+	int64_t*            D0J, *D0Si;
+	gpudata*            D1Data;
+	int64_t             D1Off;
+	int64_t*            D1J, *D1Si;
+	int64_t*            I0J, *I0Si;
+	
+	gpudata*            W;
+	int64_t             W0Off;
+	ssize_t             W1Off;
+	size_t              shmemBytes;
+	ssize_t             SHMEMK0Off;
+	ssize_t             SHMEMK1Off;
+	
+	unsigned*           perm;
+	
+	void**              kArgs;
+	
+	/* Scheduler */
+	size_t              bs;
+	size_t              gs;
+};
+typedef struct redux_ctx redux_ctx;
+
+
+
+/**
+ *                    Reduction Operator Attributes.
+ */
+
+struct GpuReductionAttr{
+	gpucontext*   gpuCtx;
+	unsigned      numProcs;
+	size_t        maxLg, maxL0, maxGg, maxG0, maxLM;
+	
+	ga_reduce_op  op;
+	int           maxSrcDims;
+	int           maxDstDims;
+	int           s0Typecode, d0Typecode, d1Typecode, i0Typecode;
+};
+
+
+/**
+ *                    Reduction Operator.
+ * 
+ * INTRO
+ * 
+ * Generates the source code for a reduction kernel over arbitrarily-ranked,
+ * -shaped and -typed tensors.
+ * 
+ * It is assumed that at most one axis will ever be of length > 2**31-1. The
+ * assumption is believed safe because no GPU or similar accelerator presently
+ * on Earth has the capacity to store or process 2**62-element tensors.
```

Comment:
@abergeron When a user broadcasts a tensor on two or more axes, each of which are of length >= 2**31, he implicitly accepts a computational cost of 2**62 FLOP's. The most powerful GPUs on Earth right now have a throughput of 1e10 `float32` additions or multiplications per second, and so would chug through such a tensor of 2**62 elements in a _mere_ 15 years. I'm unaware of a realistic environment and usecase for a 15-year computation.